File: leetcode/61.Rotate_List/cpp/main.cpp
// Definition for singly-linked list.
#include <vector>
struct ListNode {
    int val;
    ListNode* next;
    ListNode()
        : val(0)
        , next(nullptr)
    {
    }
    ListNode(int x)
        : val(x)
        , next(nullptr)
    {
    }
    ListNode(int x, ListNode* next)
        : val(x)
        , next(next)
    {
    }
};

class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k)
    {
        ListNode* dummy { head };
        ListNode* newHead {};
        int i { 0 };
        while (!dummy) {
            if (!dummy->next && i < k) {
                dummy->next = head;
            }
            if (i == k) {
                newHead = dummy->next;
                dummy   = newHead;
            } else {
                dummy = dummy->next;
            }
            i++;
        }

        return newHead;
    }
};

File: leetcode/61.Rotate_List/rust/src/main.rs
use std::{collections::VecDeque, rc::Rc};

struct Solution;
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}
impl Solution {
    pub fn rotate_right(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
        head.as_ref()?;
        let mut head = head;
        let mut deq = VecDeque::new();
        let mut current = &head;
        let mut cnt = 0;
        while let Some(node) = current {
            deq.push_back(node.val);
            current = &node.next;
            cnt += 1;
        }

        let k = k % cnt;

        for _ in 0..k {
            if let Some(val) = deq.pop_back() {
                deq.push_front(val);
            }
        }

        let mut current = head.as_mut();
        while let Some(node) = current {
            if let Some(popped) = deq.pop_front() {
                node.val = popped;
            }

            current = node.next.as_mut();
        }

        head
    }
}
